larq/optimizers_v2.py

```diff
@@ -1,10 +1,121 @@
+import warnings
+
 import tensorflow as tf
 import larq as lq
 
 from larq import utils
 from copy import deepcopy
 
 
+@utils.register_keras_custom_object
+class CaseOptimizer(tf.keras.optimizers.Optimizer):
+    """A case operation for optimizers that each train a subset of a model's variables.
+
+    An optimizer is used to train a variable iff its accompanying predicate evaluates to
+    `True`.
+
+    For each variable, at most one optimizer's predicate may evaluate to `True`. If no
+    optimizer's predicate evaluates to `True` for a variable, it is trained with the
+    `default` optimizer.
+
+    # Arguments
+    pred_opt_pairs: a list of `(tf.keras.optimizers.Optimzer, pred)` pairs, where `pred` 
+        takes one `tf.Variable` as argument and returns `True` if the optimizer should
+        train that variable, e.g. `pred(var) === True`.
+    default: a `tf.keras.optimizers.Optimizer` to be applied to any variable not claimed
+        by any other optimizer.
+    """
+
+    def __init__(self, pred_opt_pairs, default, name="optim_case"):
+        super().__init__(name=name)
+
+        for i, (pred, opt) in enumerate(pred_opt_pairs):
+            if not callable(pred):
+                raise TypeError(
+                    f"Expected callable predicate at `pred_opt_pairs[{i}][0]` but got `{type(pred)}`."
+                )
+            if not isinstance(opt, tf.keras.optimizers.Optimizer):
+                raise TypeError(
+                    f"Expected `tf.keras.optimizers.Optimizer` at `pred_opt_pairs[{i}][1]` but got `{type(opt)}`."
+                )
+        self.pred_opt_pairs = pred_opt_pairs
+
+        if default is not None and not isinstance(
+            default, tf.keras.optimizers.Optimizer
+        ):
+            raise TypeError(
+                f"Expected `tf.keras.optimizers.Optimizer` for `default` but got `{type(fallback_optimizer)}`."
+            )
+        self.default = default
+
+    def __getattr__(self, name):
+        if name == "lr":
+            raise ValueError("No single learning rate available for CaseOptimizer.")
+        return super().__getattr__(name)
+
+    def apply_gradients(self, grads_and_vars, name=None):
+        opt_grads_and_vars = [[] for _ in range(len(self.pred_opt_pairs))]
+        default_grads_and_vars = []
+
+        for grad, var in grads_and_vars:
+            num_opts = 0
+            for i, (pred, opt) in enumerate(self.pred_opt_pairs):
+                if pred(var):
+                    opt_grads_and_vars[i].append((grad, var))
+                    num_opts += 1
+
+            if num_opts == 0:
+                default_grads_and_vars.append((grad, var))
+            if num_opts > 1:
+                raise ValueError(f"Variable `{var}` claimed by multiple optimizers.")
+
+        train_ops = []
+        for (_, opt), grads_and_vars in zip(self.pred_opt_pairs, opt_grads_and_vars):
+            train_ops.append(opt.apply_gradients(grads_and_vars, name=name))
```

Comment:
I always like list comprehensions at places like this, though this might not clean up the code in this case (feel free ignore this remark)

Reply:
This is how it'd look as a list comprehension:

```python
train_ops = [
    opt.apply_gradients(grads_and_vars, name=name)
    for (_, opt), grads_and_vars in zip(self.pred_opt_pairs, opt_grads_and_vars)
]
```

I'm not 100% sure which I like better myself, wdyt?

Reply:
Another option to make it more readable, suggested by @daniel-vos:

```python
opts = [opt for (_, opt) in self.pred_opt_pairs]
train_ops = [
    opt.apply_gradients(grads_and_vars, name=name)
    for opt, grads_and_vars in zip(opts, opt_grads_and_vars)
]
```

I personally like this best; don't think the extra iteration over the optimizers matters much in terms of performance since there's not usually more than one or two optimizers.

Reply:
Changed in b33bd04.